stan/math/rev/functor/adj_jac_apply.hpp

```diff
@@ -19,53 +19,35 @@ namespace math {
 namespace internal {
 
 /**
- * Place the data for the y_adj_ into the appropriate container.
- */
-template <typename T, typename = void>
-struct build_y_adj;
-
-/**
- * Specialization returns the real adjoint value for arithmetic and eigen
- * types. Since `vari_value<Eigen>`` holds the dimensions we don't need M.
+ * Store the adjoint in y_vi[0] in y_adj
+ *
+ * @tparam size dimensionality of M
+ * @param[in] y_vi pointer to pointer to vari
+ * @param[in] M shape of y_adj
+ * @param[out] y_adj reference to variable where adjoint is to be stored
  */
 template <typename T>
-struct build_y_adj<T,
-                   require_t<disjunction<std::is_arithmetic<T>, is_eigen<T>>>> {
-  /**
-   * Retreive the adjoints for y_adj_
-   *
-   * @tparam size dimensionality of M
-   * @param[in] y_vi pointer to pointer to vari
-   * @param[in] M shape of y_adj (not used in this specialization)
-   */
-  template <typename VariType, size_t size>
-  static inline auto& apply(VariType&& y_vi, const std::array<int, size>& M) {
-    return y_vi->adj_;
-  }
-};
+inline auto build_y_adj(vari_base** y_vi, int M) {
```

Comment:
I'd recommend we don't use `vari_base` outside of vari.hpp and `chainablestack`. I might make it in namespace internal to emphasize that. I think we can actually know all of the `var_value<T>` types that come into `adj_jac`

Reply:
Cool, yeah, this part I put together in a bit of a panic at the end (just wanted something to work). I'm happy getting rid of the `dynamic_cast` s too.

Reply:
Word!